tests/testthat/test-pkg-duckdb.R

```diff
@@ -355,3 +355,16 @@ test_that("duckdb translation for n_distinct works", {
     classes = "substrait.n_distinct.approximate"
   )
 })
+
+test_that("duckdb translation for year() works", {
+  skip_if_not(has_duckdb_with_substrait())
+
+  # Error: Binder Error: Positional reference 2 out of range (total 1 columns)
```

Comment:
@paleolimbot I've tried debugging this but I'm a bit stuck.  Any ideas?

First I tried running this query a few different ways; entirely in DuckDB, then in DuckDB via Substrait (it worked), and then in DuckDB via Substrait via substrait-r (it resulted in an error):

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(duckdb)
#> Loading required package: DBI
library(substrait)
#> 
#> Attaching package: 'substrait'
#> The following object is masked from 'package:duckdb':
#> 
#>     duckdb_get_substrait
#> The following object is masked from 'package:stats':
#> 
#>     filter

# do it all in DuckDB - hooray, it works!
con <- dbConnect(duckdb::duckdb(), dbdir = ":memory:")
dbWriteTable(con, "tbl", tibble::tibble(x = as.Date("1987-10-09")))
res <- dbGetQuery(con, "SELECT EXTRACT(YEAR FROM x) FROM tbl")
print(res)
#>   main.date_part('year', x)
#> 1                      1987

# do it all in DuckDB via Substrait - hooray, it also works!
plan <- duckdb_get_substrait(
  "SELECT EXTRACT(YEAR FROM x) FROM tbl",
  tables = list(tbl = tibble::tibble(x = as.Date("1987-10-09")))
)

duckdb_from_substrait(
  plan = plan,
  tables = list(tbl = tibble::tibble(x = as.Date("1987-10-09"))),
  as_data_frame = FALSE
) %>% 
  collect()
#> # A tibble: 1 × 1
#>   `year(x)`
#>       <int>
#> 1      1987

# Use substrait-r to generate the plan and then run in DuckDB - does not work :(
out <- (
  tibble::tibble(x = as.Date("1987-10-09")) %>%
  duckdb_substrait_compiler() %>%
  substrait_project(year = year(x))
)$plan()

duckdb_from_substrait(
  plan = out,
  tables = list(named_table_1 = tibble::tibble(x = as.Date("1987-10-09"))),
  as_data_frame = FALSE) %>% 
  collect()
#> Error: Binder Error: Positional reference 2 out of range (total 1 columns)
```

I compared the two plans, but they seem pretty similar.  I guess I'll check what `maintain_singular_struct` is?

Reply:
So we'd have to implement something here which is more complex; some attribute or parameter to implement in `substrait_call()`, which when set to a then would be evaluated by `substrait_project()` and then do whatever is needed to set `maintain_singular_struct` to `true`.

Reply:
Given that the direct running of extra year creates a column called `main.date_part('year', x)`, I wonder if this is a clue to the fact that it's been mapped to an internal function that returns multiple items?

Reply:
Ah, that's a reasonable point actually; the naming is not canonical, and they don't actually implement the canonical version at all, this is just an equivalent function.  It should be `extract()` with an option set to `year`, so maybe I shouldn't be mapping against this one (literally called `year`, which doesn't exist in Substrait) at all?